### DISCUSSION TITLE
[Week21] React 中改變二元陣列中的數值

### DISCUSSION THREAD

**Discussion by pcchen95** [Lidemy/forum#200] [前端相關問題] [answered]
大家好，

下面是我擷取作業二 - 五子棋的程式碼部分內容，這個 function 是用來 render 同一個 row id 的所有 component `Square`，並且點擊會觸發 `handleClickSquare` 去修改記錄每個位置內容的二元陣列，將該位置狀態設定為 `1`。

```jsx
function BoardRow({ row }) {
  // value 為一個物件，存放一個名為 squares 的 19*19 二元陣列
  const [value, setValue] = useContext(valueContext);

  const handleClickSquare = (e, [y, x]) => {
    let squares = value.squares.slice();
    squares[y][x] = "1";
    // console.log(squares);
    setValue({ squares });
  };

  let rows = [];
  for (let i = 0; i < size; i++) {
    rows[i] = (
      <Square key={Math.random()} onClick={(e) => handleClickSquare(e, [row, i])} >
      </Square>
    );
  }
  return <BoardColumn>{rows}</BoardColumn>;
}
```

`value.squares` 是一個二元陣列，預設值都是 null，而 `squares` 是一個和 `value.squares` 內容一模一樣的二元陣列。

這裡假設我嘗試點選的是第 2 個 column 的第 1 個 `Square`，也就是程式碼裡 `y = 1`、`x = 0`。透過 `squares[y][x] = "1"` 去變更陣列內容後，預期應該只有 `sqaures[1][0]` 這個位置被改成 `1`，console.log 的結果應該如下：
```
[
  [null, null, null, ... , null],
  ['1', null, null, ... , null],
  [null, null, null, ... , null],
  .
  .
  .
  [null, null, null, ... , null],
]
```

但實際卻是`每個 row 的第一個位置都被變更了`：
```
[
  ['1', null, null, ... , null],
  ['1', null, null, ... , null],
  ['1', null, null, ... , null],
  .
  .
  .
  ['1', null, null, ... , null],
]
```

研究了一陣子找不到問題點所在，請再幫忙回答，謝謝！

**Comment by aszx87410** [comment, marked as answer]
我猜測你的 squares 是這樣產生的：`Array(19).fill(Array(19).fill(null))`
這樣寫其實就是這樣：

``` js
var arr = Array(19).fill(null)
var arr2 = Array(19).fill(arr)
```

所以 arr2 的每一個元素都會是 arr，因此你改變 arr，看起來當然 arr2 裡的每一列都會改變

**Reply by aszx87410** [reply]
其他同學也有碰過，可參考：https://github.com/Lidemy/mentor-program-5th-rockyooooooo/pull/34

**Reply by pcchen95** [reply]
是用和老師寫的一樣的方法產生的沒錯（太神啦～）！
完全沒想到錯誤是出在這邊，因為用 fill 的話裡面每次放進去的 arr 都指向同一個記憶體位置，所以改了一個就會影響到全部。
改成雙重迴圈就馬上解決問題了，感謝老師～
